lib/sycamore/sycamore/connectors/common.py

```diff
@@ -88,49 +112,86 @@ def flatten_data(
     return items
 
 
-def unflatten_data(data: dict[str, Any], separator: str = ".") -> dict[Any, Any]:
-    result: dict[Any, Any] = {}
+def unflatten_data(data: dict[Any, Any], separator: str = ".") -> dict[Any, Any]:
+    """
+    Unflattens a dictionary with keys that contain separators into a nested dictionary. The separator can be escaped,
+    and if there are integer keys in the path, the result will be a list instead of a dictionary.
+    """
 
-    def parse_key(key: str) -> list:
-        # Handle escaped separator
+    def split_key(key: str, separator: str = ".") -> list[str]:
+        """
+        Splits the key by separator (which can be multiple characters), respecting escaped separators.
+        """
         parts = []
         current = ""
-        escape = False
-        for char in key:
-            if escape:
-                if char == separator:
-                    current += separator
+        i = 0
+        while i < len(key):
+            if key[i] == "\\":
+                # Escape character
+                if i + 1 < len(key):
+                    current += key[i + 1]
+                    i += 2
                 else:
-                    current += "\\" + char
-                escape = False
-            elif char == "\\":
-                escape = True
-            elif char == separator:
+                    # Trailing backslash, treat it as literal backslash
+                    current += "\\"
+                    i += 1
+            elif i + len(separator) < len(key) and key[i : i + len(separator)] == separator:
+                # Found separator
                 parts.append(current)
                 current = ""
+                i += len(separator)
             else:
-                current += char
+                current += key[i]
+                i += 1
         parts.append(current)
         return parts
 
-    for key, value in data.items():
-        parts = parse_key(key)
+    result: dict[Any, Any] = {}
+    for flat_key, value in data.items():
+        parts = split_key(flat_key, separator)
         current = result
         for i, part in enumerate(parts):
-            part_key: Union[str, int] = int(part) if part.isdigit() else part
+            # Determine whether the key part is an integer (for list indices)
+            key: Union[str, int]
+            try:
+                key = int(part)
+            except ValueError:
+                key = part
```

Comment:
could prob do `if key.isnumeric()`

Reply:
Part will also be a `str` though? So I would have to cast it to make it work I think.

Reply:
isnumeric is a string method to tell you if all the characters are digits. so you could do
```python
if part.isnumeric():
    part = int(part)
```

Reply:
Ah that makes sense. I can do that.

Reply:
wait isdigit does the same thing. Why did we change this from the prev implementation?

Reply:
This is because both `isdigit` and `isnumeric` don't handle decimals or negative numbers. I'll stick with the current implementation for now.